feat(RELEASE-1247): add idempotent behavior to publish-to-mrrc task

elenagerman opened the pull request:
## Describe your changes
Implements idempotency for the publish-to-mrrc task to prevent redundant uploads
and resource waste when the task is retried or re-run with the same inputs.

HLD: https://docs.google.com/document/d/16FwCQD7tDjZJwD2YE3tlzhwr_-afM6grv95q7ijrbBo/edit?tab=t.0

## Relevant Jira
[RELEASE-1247](https://issues.redhat.com/browse/RELEASE-1247)

## Checklist before requesting a review
- [x] I have marked as draft or added `do not merge` label if there's a dependency PR
  - If you want reviews on your draft PR, you can add reviewers or add the `release-service-maintainers` handle if you are unsure who to tag
- [x] My commit message includes `Signed-off-by: My name <email>`
- [x] I read CONTRIBUTING.MD and [commit formatting](CONTRIBUTING.md#commit-message-formatting-and-standards)
- [x] I have run the README.md generator script in `.github/scripts/readme_generator.sh` and verified the results using `.github/scripts/check_readme.sh`

elenagerman commented:
/retest

elenagerman commented:
>      * Are these marker files persistent for usage between different pipelineruns?

yes, they are persistent. it stored under 
`
        # Create idempotency marker directory
        MARKER_DIR="$(params.dataDir)/.mrrc-published"
`

johnbieren commented:
> @ligangty @johnbieren As an option, we could also consider a manual cleanup via a separate task (add a new task or parameter to delete specific marker files before re-running the pipeline or enhance the check to verify the artifacts still exist in mrrc before trusting the marker file. And the simplest, but a rather crude solution (IMO, it doesn't directly solve the problem) is to add timestamps to marker files and auto-expire after a certain period WDYT?

I don't really like the separate task option

elenagerman commented:
> I mean, when would that be passed? I have a release, it passed. I manually rollback the image as Gang said. I rerun the release. Wouldn't the parameters be the same to the task each time?
@johnbieren 
Makes sense, my bad, I missed it.
In this case, a separate-task solution would be ideal.
But, agree, the con is that we'll have an extra task to maintain.
To keep it simple, I integrated a 'self-cleaning' step into the publish-to-mrrc task instead. 
It's triggered by the mrrc.republish annotation.
Please let me know WDYT

johnbieren commented:
> > I mean, when would that be passed? I have a release, it passed. I manually rollback the image as Gang said. I rerun the release. Wouldn't the parameters be the same to the task each time?
> 
> @johnbieren Makes sense, my bad, I missed it. In this case, a separate-task solution would be ideal. But, agree, the con is that we'll have an extra task to maintain. To keep it simple, I integrated a 'self-cleaning' step into the publish-to-mrrc task instead. It's triggered by the mrrc.republish annotation. Please let me know WDYT

I'd defer to @ligangty

ligangty commented on tasks/managed/publish-to-mrrc/publish-to-mrrc.yaml:
```diff
@@ -175,6 +190,92 @@ spec:
+            MARKER_DIR="$(params.dataDir)/.mrrc-published"
+            if [[ -d "$MARKER_DIR" ]]; then
+              rm -rf "$MARKER_DIR"
+              echo "Removed marker directory: $MARKER_DIR"
```
I cannot confirm if this makes sense. Should we remove the whole directory which will cause the removal of other marker files not belonging to current release?

elenagerman replied:
Removing the entire directory is SAFE.
This removes only the markers belonging to the current release.
The dataDir is specific to THIS release (populated from its own Trusted Artifacts), and other releases have a completely separate dataDir and Trusted Artifacts. Hence, no other release's markers exist in this directory.

elenagerman commented:
> I really don't like having the user change something to republish from what they had when they published. I guess to get at the main point of [RELEASE-1247](https://issues.redhat.com/browse/RELEASE-1247), what happens if we run this task on the same artifact multiple times? Does it _do_ anything after the first run? If yes, how do we properly fix that?

The problem is that after a charon rollback, all artifacts are deleted from mrrc, but marker files still exist in trusted artifacts.
Rerunning the release → task sees markers → skips → nothing gets published 
Users cannot just rerun with the same parameters, because that's what idempotency prevents! 
When we rerun a release, you cannot change parameters (they come from RPA). So we need some external signal to say "ignore idempotency markers this time."
The annotation is a feature, not a bug - it's an explicit, safe mechanism to override idempotency only when actually needed (rollback scenarios).

Another possible option to solve the rollback issue is to automatically detect a charon rollback by checking whether mrrc still has the artifacts. But it will be much heavier on performance, since we'll have an HTTP call to mrrc on every run to verify artifacts.
In addition, in this case, we can't distinguish between charon rollback or manual artifact deletion for other reasons or temporary mrrc issues, and could accidentally republish when not needed

johnbieren commented:
> > I really don't like having the user change something to republish from what they had when they published. I guess to get at the main point of [RELEASE-1247](https://issues.redhat.com/browse/RELEASE-1247), what happens if we run this task on the same artifact multiple times? Does it _do_ anything after the first run? If yes, how do we properly fix that?
> 
> The problem is that after a charon rollback, all artifacts are deleted from mrrc, but marker files still exist in trusted artifacts. Rerunning the release → task sees markers → skips → nothing gets published Users cannot just rerun with the same parameters, because that's what idempotency prevents! When we rerun a release, you cannot change parameters (they come from RPA). So we need some external signal to say "ignore idempotency markers this time." The annotation is a feature, not a bug - it's an explicit, safe mechanism to override idempotency only when actually needed (rollback scenarios).
> 
> Another possible option to solve the rollback issue is to automatically detect a charon rollback by checking whether mrrc still has the artifacts. But it will be much heavier on performance, since we'll have an HTTP call to mrrc on every run to verify artifacts. In addition, in this case, we can't distinguish between charon rollback or manual artifact deletion for other reasons or temporary mrrc issues, and could accidentally republish when not needed

I'm not following all of this.
> but marker files still exist in trusted artifacts.

trusted artifacts are per taskRun. Separate releases don't reuse the same trusted artifacts.

> So we need some external signal to say "ignore idempotency markers this time."

I don't think this is a good pattern. The whole goal of the idempotent work is so that people can run the release pipeline one time or twenty times with the same input, and it is only added to its destination once. I don't think we should be introducing "ignore idempotent" options

> But it will be much heavier on performance, since we'll have an HTTP call to mrrc on every run to verify artifacts

We check the existing digest for push-snapshot to make sure we don't push if it is already there. How is this different?

> In addition, in this case, we can't distinguish between charon rollback or manual artifact deletion for other reasons or temporary mrrc issues, and could accidentally republish when not needed

Huh? I don't think it matters how the artifact was removed. If the artifact is not there and the pipeline runs, it should be put there. The only "republish when not needed" to be worried about is if the exact artifact is already there

elenagerman commented:
> > @johnbieren @ligangty I see that our discussion for rollback case coverage has grown too large. What do you think about excluding this part from this commit and opening a separate Jira&PR? At the same time, I'll create a separate document to discuss a possible solution.
> 
> The existing ticket is for making the task idempotent. I don't understand why we'd open a second jira for it. All for a document about a solution though

I didn't mean to open another Jira for the idempotent feature. I'm asking only about the rollback case support part.
But if you prefer to have it in the same PR, let's continue our discussion :)

elenagerman commented:
> We check the existing digest for push-snapshot to make sure we don't push if it is already there. How is this different?

Correct me if I'm wrong, but charon doesn't have a standalone **check** or **exists** command. Instead, it relies on the marker files in the dataDir to handle idempotency. 
The alternative, like a manual curl to the mrrc API, is complex, error-prone, and has no standard charon interface.
Unlike push-snapshot, which provides OCI registry APIs for querying digests, mrrc/charon doesn't provide this capability out of the box.

ligangty commented on tasks/managed/publish-to-mrrc/publish-to-mrrc.yaml:
```diff
@@ -175,6 +218,120 @@ spec:
+          # Construct potential MRRC URL (may need adjustment based on actual MRRC structure)
+          MRRC_BASE_URL="https://maven.repository.redhat.com"
+          # Try to query the product/version in the target repository
+          MRRC_CHECK_URL="${MRRC_BASE_URL}/${target_repo}/${product}/${version}/"
```
MRRC is not organized by this structure. It is a standard maven repository which is organized by $groupId/$artifactId/$version (GAV), and this version is not aligned with product version. 
A product version with the maven zip is made up by many GAVs. Actually, MRRC does not provide a way to check if a product+version for the maven zip is published.

elenagerman replied:
Not good :(
Thanks a lot for the clarification @ligangty !
But since mrrc is organized by maven GAV and a product version contains many GAVs, we cannot query mrrc to check if a product+version has been published. 
Because in this case, we'd need to parse the maven zip to extract all GAVs, query each GAV individually in mrrc, and only then determine if all are present
But this would be extremely slow (hundreds of HTTP requests) and absolutely unreliable (partial publishes, timing issues)

ligangty commented:
I'm thinking if we really need this idempotent feature for the MRRC, because currently charon allows a certain level of the re-publishing. It does not override the content if they are already there, instead it will give a metadata to notify that "there is a re-publishing happened for this GAV" for future audit. In other words, we don't forbid this type of re-publishing.
I don't know if this is a feature from Konflux perspective. If it is, let's continue on it.

elenagerman commented:
> This makes it sound like charon already has idempotent behavior baked in so we may need no action on our side. I am good with that (and closing this PR + jira) if that is the case.
> We would have to document it in the [release-to-mrrc](https://issues.redhat.com/browse/RELEASE-to-mrrc) pipeline README though. Something along the lines of "charon has no mechanism for detecting that a maven zip was already published. Thus, no artifacts will be filtered out and all will be passed to the conforma check and other release tasks"

Thank you @johnbieren ! 
I agree with reverting the changes for now. 
However, I was thinking—instead of closing this PR, could we keep it open and use it just for the README update? 
I've already reverted the idempotency code and added the documentation there instead. 
PTAL

johnbieren commented on pipelines/managed/release-to-mrrc/README.md:
```diff
@@ -3,6 +3,13 @@
+
+Charon (the MRRC publishing tool) has no mechanism for detecting that a maven zip was already published. Thus, when
+rerunning this pipeline with the same release, no artifacts will be filtered out and all will be passed to the
+conformance check and other release tasks. This means the pipeline will re-process all artifacts on every run, but
```
`s/conformance/conforma/`

elenagerman replied:
sorry, missed this change was wrongly done by my Grammarly speller :) 
fixed

elenagerman committed:
docs(RELEASE-1247): document Charon idempotency in MRRC pipeline

Add documentation note explaining that Charon (MRRC publishing tool)
has no mechanism for detecting already-published maven zips. This
clarifies that:

- When rerunning the pipeline with the same release, all artifacts
  will be reprocessed and passed to conforma check
- No artifact filtering occurs at the pipeline level
- Charon handles duplicate publications at the MRRC level

This addresses questions raised in RELEASE-1247 about whether
idempotency support was needed. Since Charon already handles
duplicates at the MRRC level, no additional implementation is
required in the pipeline.

Signed-off-by: Elena German <elgerman@redhat.com>
